feat(connect-cardano): add support for babbage features

gabrielKerekes opened the pull request.

szymonlesisz committed:
fix(suite): preload store before first render

szymonlesisz committed:
chore(suite): remove redunant "storageLoaded" field

storage is always loaded now

szymonlesisz committed:
feat(suite-desktop-api): add handshake method

szymonlesisz committed:
feat(suite-desktop): implement desktopApi handshake

mroz22 committed:
feat(auth-server): create auth server

gabrielKerekes committed:
feat(connect-cardano): add support for babbage features

mroz22 committed:
feat(suite): re-enable implicit oauth flow

davidmisiak commented on packages/connect/src/api/cardanoSignTransaction.ts:
```diff
@@ -265,6 +289,24 @@ export default class CardanoSignTransaction extends AbstractMethod<
             this._ensureFeatureIsSupported('Plutus');
         }
+
+        params.outputsWithData.forEach(({ output, inlineDatum, referenceScript }) => {
```
What about merging this with the `params.outputsWithData.forEach(({ output })...` above?

gabrielKerekes replied:
I prefer having the checks separated. I like the fact that the checks are chronological as they were added and it's IMHO clearer what's supported with each version.

davidmisiak replied:
Alright.

davidmisiak commented on packages/connect/src/api/cardano/cardanoUtils.ts:
```diff
@@ -0,0 +1,17 @@
+    while (processedSize < data.length) {
+        const chunk = data.slice(processedSize, processedSize + chunkSize);
+        await typedCall(messageType, 'CardanoTxItemAck', {
+            data: chunk,
```
How does this work? `typedCall` expects a hex-encoded string when sending bytes?

gabrielKerekes replied:
I think so, yes. Similarly to other `bytes` types e.g. `CardanoTxInput.prev_hash` or `CardanoTxOutput.datum_hash`. I think the protobuf layer somehow deals with the conversion to bytes, but I haven't checked exactly. I just know that this worked.

Even when a `Buffer` is used as an incoming type into Connect it gets converted to hex string before being sent to Trezor: https://github.com/trezor/trezor-suite/blob/develop/packages/connect/src/api/bitcoin/inputs.ts#L73

davidmisiak replied:
Ok, thanks.

davidmisiak commented on packages/connect/src/api/cardano/cardanoOutputs.ts:
```diff
@@ -43,3 +57,40 @@ export const transformOutput = (output: any): OutputWithTokens => {
 };
+
+export const sendOutput = async (typedCall: any, outputWithData: OutputWithData) => {
+    const MAX_CHUNK_SIZE = 512;
```
I don't like this number being different than the number in FW (even though it represents the same number of bytes). I would prefer converting the data from hex-encoded string to bytes asap and dealing only with bytes later on - is that possible?

gabrielKerekes replied:
I wouldn't convert to bytes before sending, see [comment](https://github.com/vacuumlabs/trezor-suite/pull/1#discussion_r907136861) above.

Perhaps this would be better?
```suggestion
    const MAX_CHUNK_SIZE = 256 * 2; // 256 hex-encoded bytes
```

gabrielKerekes replied:
I added my suggestion in https://github.com/vacuumlabs/trezor-suite/pull/1/commits/58062096a0b3ab7b5e023ad1e11789368ceba375.

davidmisiak replied:
Alright, your suggestion seems reasonable to me 👍

davidmisiak commented on packages/connect/src/api/cardano/cardanoOutputs.ts:
```diff
@@ -1,30 +1,44 @@
 // origin: https://github.com/trezor/connect/blob/develop/src/js/core/methods/helpers/cardanoOutputs.js
 
+/* eslint-disable no-restricted-syntax */
```
Just out of curiosity, why is this needed?

gabrielKerekes replied:
Because of the `for (const assetGroup of tokenBundle)` syntax. Eslint doesn't like that and I suppose it prefers `await Promise.awaitAll(tokenBundle.map(async ...))` (I didn't check the syntax is correct) but I think the order isn't guaranteed in that case so it wouldn't work for us. And AFAIK async foreach doesn't exist. The same is used in `cardanoSignTransaction.ts` btw.

I'll add a comment here (and in `cardanoSignTransaction.ts`) to make it clear.

gabrielKerekes replied:
Done in https://github.com/vacuumlabs/trezor-suite/pull/1/commits/58062096a0b3ab7b5e023ad1e11789368ceba375.

davidmisiak commented on packages/connect/src/types/api/cardanoSignTransaction.ts:
```diff
@@ -19,19 +19,21 @@ export interface CardanoAssetGroup {
 }
 
-export type CardanoOutput =
+export type CardanoOutput = (
```
Hm, it seems like they removed Flow types (finally!), is that right? In that case, we should update/remove the places where we reference them, such as `cardanoSignTransaction.md` and other `.md` files.

Also, `cardanoSignTransaction.md` should be updated with the new Babbage tx items.

gabrielKerekes replied:
Yes, indeed. Good points, I'll update the `.md` files, thanks 👍

gabrielKerekes replied:
Done in https://github.com/vacuumlabs/trezor-suite/pull/1/commits/6ef0d71261ccbf7424f20d36d7eef7635cfefab1.

davidmisiak replied:
Cool, but shouldn't we also add the descriptions of the new Babbage tx items to `cardanoSignTransaction.md`?

gabrielKerekes replied:
I think we should :D Sorry, I completely missed it. Updated. I squashed the commit though.

davidmisiak commented on packages/integration-tests/projects/connect/__fixtures__/cardanoSignTransaction.ts:
```diff
@@ -347,6 +399,12 @@ const legacyResults = {
         payload: false,
     },
+    beforeBabbage: {
+        // older FW doesn't support babbage-related features (inlineDatum, referenceScript, collateral return,
```
Nit: `collateral return` has different formatting.

gabrielKerekes replied:
👍 Fixed in https://github.com/vacuumlabs/trezor-suite/pull/1/commits/58062096a0b3ab7b5e023ad1e11789368ceba375.

davidmisiak commented on packages/integration-tests/projects/connect/__fixtures__/cardanoSignTransaction.ts:
```diff
@@ -347,6 +399,12 @@ const legacyResults = {
+        // totalCollateral, referenceInputs)
+        rules: ['<2.5.2', '1'],
+        payload: false,
+    },
```
Reminder: We should update `multisigWithMostElementsFilledAndSharedWithLedger` (at least add a Babbage-formatted output, or maybe create a similar testcase for Alonzo&Babbage items?).

gabrielKerekes replied:
I added such a test - https://github.com/vacuumlabs/trezor-suite/pull/1/files#diff-6e989ad5e80587d676b0177bf290b661fc11290f47f3348fa1b1f16b6f9ed0e1R2398.

davidmisiak commented on packages/integration-tests/projects/connect/__fixtures__/cardanoSignTransaction.ts:
```diff
@@ -2101,5 +2166,342 @@ export default {
+                outputs: [
+                    SAMPLE_OUTPUTS.simple_shelley_output,
+                    SAMPLE_OUTPUTS.output_with_datum_hash_and_base_address_paramaters,
+                ],
```
I'm afraid both outputs use the legacy format (despite the testcase name).

gabrielKerekes replied:
Nice, I'm not sure how that happened. Thanks! Fixed in https://github.com/vacuumlabs/trezor-suite/pull/1/commits/58062096a0b3ab7b5e023ad1e11789368ceba375

davidmisiak reviewed (commented):
A few questions and suggestions below. I haven't run anything since I don't have the Suite set up.

gabrielKerekes committed:
chore(connect-cardano): update docs with Babbage and TypeScript

komret committed:
feat(suite): enable both oauth flows

komret committed:
feat(suite): improve oauth token refresh

komret committed:
feat(suite): remove access_token from storage, remove google-oauth-library

mroz22 committed:
test(e2e): test bridge in desktop

bouchja1 committed:
feat(suite-native): Fix trezor/connect RN network utils

komret committed:
fix(components): set Select width to 100%

marekrjpolak committed:
chore(suite-desktop): modules rework

marekrjpolak committed:
chore(suite-desktop): app starting cleanup

komret committed:
feat(suite): store both refesh and access token for Google OAuth

marekrjpolak committed:
refactor(suite-desktop): unify entrypoint with web

marekrjpolak committed:
refactor(suite): handleProtocolRequest thunk

marekrjpolak committed:
chore: remove deprecated events

komret committed:
fix(suite): oauth metadata fetching, comments

Nodonisko committed:
chore(native): add redux to app, basic app setings reducer (#5716)

komret committed:
feat(suite): tweak token migration and update README

marekrjpolak committed:
feat(suite-desktop): user-data/get-info handshake

marekrjpolak committed:
feat(suite-desktop): server/request-address handshake

dependabot[bot] committed:
chore(deps): bump parse-url from 6.0.0 to 6.0.2

Bumps [parse-url](https://github.com/IonicaBizau/parse-url) from 6.0.0 to 6.0.2.
- [Release notes](https://github.com/IonicaBizau/parse-url/releases)
- [Commits](https://github.com/IonicaBizau/parse-url/commits)

---
updated-dependencies:
- dependency-name: parse-url
  dependency-type: indirect
...

Signed-off-by: dependabot[bot] <support@github.com>

marekrjpolak committed:
feat(suite-desktop): firmware baseUrl in handshake

bouchja1 committed:
feat(websocket): Add ws interface for React Native with user agent

matejkriz committed:
chore(storage): fix unexpected block statement lint error

matejkriz committed:
chore(scripts): fix eslint warning

console log is okay in this script

komret committed:
feat(auth-server): specify allowed origins

komret committed:
feat(auth-server): add production credentials

komret committed:
fix(suite): consolidate OAuth credentials and URLs

komret committed:
e2e(suite): update fixtures to new tokens storage

mroz22 committed:
feat(connect): add getFirmwareHash method

mroz22 committed:
test: getFirmwareHash tests

mroz22 committed:
feat(suite): implement TrezorConnect.getFirmwareHash

komret committed:
feat(docs): update metadata docs

bouchja1 committed:
feat(atom): Bottom sheet modal animation (#5715)

* feat(atom): Bottom sheet modal animation

mroz22 committed:
ci(gh): fix loading available fws from tenv in connect tests

matejkriz committed:
fix(auto-update): fix import of openpgp that was undefined after upgrade

marekrjpolak committed:
feat(suite): app init UI (both web & desktop)

mroz22 committed:
test(suite-web): fix loading available fws

bouchja1 committed:
feat(ci): Add github action workflow for develop ios build (#5724)

sime committed:
chore(docs): Migrate wiki.trezor.io Trezor Connect API doc

marekrjpolak committed:
test(suite): desktop reducer unit test

bouchja1 committed:
fix(ci): RN mac os develop build, add node from .nvmrc (#5761)

mroz22 committed:
feat(analytics): add firmware validate hash events

komret committed:
fix(suite-desktop): add dropbox to allowed origins

komret committed:
chore(suite): add oauth description and clean up last migration

mroz22 committed:
feat(fw-update): dont validate hash for custom fw

bouchja1 committed:
fix(ci): Build libs before build (#5766)

matejkriz committed:
fix(auto-update): patch GitHub provider for electron-updater to make it backward compatible with pre-release withou channels

Changes in electron-updater library broke our Github flow with pre-release status without channels (-beta/-alba). We are not able to start using -beta tags without publishing two separate builds with different versions (one for EAP, second for general public release). This patch basically reverts the changes for apps without channel tag in version.

marekrjpolak committed:
fix(suite): redux init with dbError

mroz22 committed:
test(connect): add missing binance legacy results

juriczech committed:
feat(eslint): Import ordering (#5751)

komret committed:
fix(suite): change auth server url

matejkriz committed:
chore(suite): bump beta version to 22.8.0

tomasklim committed:
fix(suite): registerProtocolHandler in old browsers

szymonlesisz committed:
fix(suite): regtest segwit addresses support

Nodonisko committed:
fix: unify react-redux version to 7.2.2

mroz22 committed:
test(connect-popup): fix cardano sign test with latest fw

marekrjpolak committed:
fix(suite): analyticsReducer init from storage

LukasRada committed:
fix(coinmarket): fix overview savings sum

marekrjpolak committed:
fix(suite): coinmarketReducer init from storage

bouchja1 committed:
fix(ci) build libs for RN project (#5779)

komret committed:
fix(auth-server): change allowed origins

komret committed:
feat(suite): enable multiple oauth server urls

marekrjpolak committed:
test(suite): unit test STORAGE.LOAD fix

mroz22 committed:
ci: run emu from url

juriczech committed:
feat(suite-native): Transaction Item (#5783)

marekrjpolak committed:
refactor(suite): suite init polished

szymonlesisz committed:
Fix/transaction utils (#5785)

tomasklim committed:
fix(suite): error on favicon change on os theme change

mroz22 committed:
chore(fw-update): revamp error message displayed to user

komret committed:
fix(suite): disable Google OAuth code flow for web

sime committed:
chore(docs): Manifest errors linking to old connect repo

mroz22 committed:
ci: migrate blockchain-link integration to github

juriczech committed:
feat(suite-native): Asset item (#5806)

matejkriz committed:
fix(sentry): smarter regex for redacting user path

- allow more slashes
- stop on special characters

vdovhanych committed:
chore(ci): adjust yarn install on suite-native workflow

davidmisiak commented:
Merged in https://github.com/trezor/trezor-suite/pull/5752
